### DISCUSSION TITLE
CAN message ID bit 24

### DISCUSSION THREAD

**Discussion by cdyellick** [canboat/canboat#254] [Q&A] [answered]
What started as an exercise in troubleshooting my NMEA 2000 network has led me down the rabbit hole of trying to understand the NMEA 2000 protocol.  In the process, I came across the CANboat project, which I have to say is an impressive corpus of work.

I’m currently trying to understand the CAN message ID.  Most of it seems straight forward, but I’m having trouble understanding the significance of bit 24.

The 1999 article, “NMEA 2000 Explained - The Latest Word,” describes the CAN identification field.  https://www.nmea.org/Assets/2000-explained-white-paper.pdf page 11.  The description indicates that bits 24 and 25 are reserved for future use.  However, the CANboat code appears to use bit 24 as the left most bit of the PNG.  This use is shown in the 2010 blog post,  http://yachtelectronics.blogspot.com/2010/09/code-for-converting-iso-11783-id-to.html, where bit 24 is stored in DP.

DP is set as 

`unsigned char DP = (unsigned char) (id >> 24) & 1;`

The PNG is defined as  

`*pgn = (DP << 16) + (PF << 8);`

or

`*pgn = (DP << 16) + (PF << 8) + PS;`

depending on if it’s PDU1 or PDU2

I realize that the 1999 NMEA article is probably out of date, but its not clear to me what bit 24 is encoding.  Did they simply decide they needed more PGNs and so added a bit?  Or is there something more to it?

**Comment by keesverruijt** [comment, marked as answer]
Hi @cdyellick , interestingly we just changed this -- see https://github.com/canboat/canboat/issues/248 . This has added another "reserved" bit to the PGN decoding.

As to the reasoning made by closed standards committees, your guess is as good as mine.

**Reply by cdyellick** [reply]
Interesting and very helpful.  Thank you for the quick reply and the great project.
